File: src/util/entities/Sticker.ts
/*
	Spacebar: A FOSS re-implementation and extension of the Discord.com backend.
	Copyright (C) 2023 Spacebar and Spacebar Contributors

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU Affero General Public License as published
	by the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU Affero General Public License for more details.

	You should have received a copy of the GNU Affero General Public License
	along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

import { Column, Entity, JoinColumn, ManyToOne, RelationId } from "typeorm";
import { StickerFormatType, StickerType } from "@spacebar/schemas";
import { BaseClass } from "./BaseClass";
import type { Guild } from "./Guild";
import type { User } from "./User";
import type { StickerPack } from "./StickerPack";

@Entity({
    name: "stickers",
})
export class Sticker extends BaseClass {
    @Column()
    name: string;

    @Column({ nullable: true })
    description?: string;

    @Column({ nullable: true })
    available?: boolean;

    @Column({ nullable: true })
    tags?: string;

    @Column({ nullable: true })
    @RelationId((sticker: Sticker) => sticker.pack)
    pack_id?: string;

    @JoinColumn({ name: "pack_id" })
    @ManyToOne(() => require("./StickerPack").StickerPack, {
        onDelete: "CASCADE",
        nullable: true,
    })
    pack: StickerPack;

    @Column({ nullable: true })
    guild_id?: string;

    @JoinColumn({ name: "guild_id" })
    @ManyToOne(() => require("./Guild").Guild, (guild: Guild) => guild.stickers, {
        onDelete: "CASCADE",
    })
    guild?: Guild;

    @Column({ nullable: true })
    user_id?: string;

    @JoinColumn({ name: "user_id" })
    @ManyToOne(() => require("./User").User, {
        onDelete: "CASCADE",
    })
    user?: User;

    @Column({ type: "int" })
    type: StickerType;

    @Column({ type: "int" })
    format_type: StickerFormatType;
}

File: extra/admin-api/Spacebar.Db/Models/Sticker.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;

namespace Spacebar.Db.Models;

[Table("stickers")]
public partial class Sticker
{
    [Key]
    [Column("id", TypeName = "character varying")]
    public string Id { get; set; } = null!;

    [Column("name", TypeName = "character varying")]
    public string Name { get; set; } = null!;

    [Column("description", TypeName = "character varying")]
    public string? Description { get; set; }

    [Column("available")]
    public bool? Available { get; set; }

    [Column("tags", TypeName = "character varying")]
    public string? Tags { get; set; }

    [Column("pack_id", TypeName = "character varying")]
    public string? PackId { get; set; }

    [Column("guild_id", TypeName = "character varying")]
    public string? GuildId { get; set; }

    [Column("user_id", TypeName = "character varying")]
    public string? UserId { get; set; }

    [Column("type")]
    public int Type { get; set; }

    [Column("format_type")]
    public int FormatType { get; set; }

    [ForeignKey("GuildId")]
    [InverseProperty("Stickers")]
    public virtual Guild? Guild { get; set; }

    [ForeignKey("PackId")]
    [InverseProperty("Stickers")]
    public virtual StickerPack? Pack { get; set; }

    [InverseProperty("CoverStickerId1Navigation")]
    public virtual ICollection<StickerPack> StickerPacks { get; set; } = new List<StickerPack>();

    [ForeignKey("UserId")]
    [InverseProperty("Stickers")]
    public virtual User? User { get; set; }

    [ForeignKey("StickersId")]
    [InverseProperty("Stickers")]
    public virtual ICollection<Message> Messages { get; set; } = new List<Message>();
}
